draft-js-plugins-editor/src/Editor/index.js

```diff
@@ -74,62 +92,88 @@ class PluginEditor extends Component {
     this.onChange(moveSelectionToEnd(editorState));
   }
 
-  componentWillReceiveProps(next) {
+  componentDidUpdate(prev: EditorProps) {
     const curr = this.props;
-    const currDec = curr.editorState.getDecorator();
-    const nextDec = next.editorState.getDecorator();
 
-    // If there is not current decorator, there's nothing to carry over to the next editor state
-    if (!currDec) return;
-    // If the current decorator is the same as the new one, don't call onChange to avoid infinite loops
-    if (currDec === nextDec) return;
-    // If the old and the new decorator are the same, but no the same object, also don't call onChange to avoid infinite loops
-    if (currDec && nextDec && getDecoratorLength(currDec) === getDecoratorLength(nextDec)) return;
+    // check if props.plugins or props.decorators are still the same, only then reset
+    if (!shallowEqual(curr.plugins, prev.plugins) ||
+        !shallowEqual(curr.decorators, prev.decorators)
+    ) {
+      const editorState = EditorState.set(
+        curr.editorState,
+        { decorator: resolveDecorators(curr, this.getEditorState, this.onChange) }
+      );
 
-    const editorState = EditorState.set(next.editorState, { decorator: currDec });
-    this.onChange(moveSelectionToEnd(editorState));
+      this.onChange(editorState);
+      this.setupPlugins();
+    }
   }
 
   componentWillUnmount() {
-    this.resolvePlugins().forEach((plugin) => {
-      if (plugin.willUnmount) {
-        plugin.willUnmount({
-          getEditorState: this.getEditorState,
-          setEditorState: this.onChange,
-        });
+    const pluginMethods = this.getPluginMethods();
+    [this.props, ...this.resolvePlugins()].forEach((plugin) => {
+      if (plugin.willUnmount != null) {
+        plugin.willUnmount(pluginMethods);
       }
     });
   }
 
   // Cycle through the plugins, changing the editor state with what the plugins
   // changed (or didn't)
-  onChange = (editorState) => {
+  onChange = (editorState: EditorState) => {
     let newEditorState = editorState;
+    const pluginMethods = this.getPluginMethods();
     this.resolvePlugins().forEach((plugin) => {
-      if (plugin.onChange) {
-        newEditorState = plugin.onChange(newEditorState, this.getPluginMethods());
+      if (plugin.onChange != null && newEditorState != null) {
```

Comment:
so if I call onChange with `null` it will not trigger the onChange callbacks for plugins. Is this desired behaviour?